processor/src/advice/providers.rs

```diff
@@ -443,3 +429,38 @@ impl AdviceProvider for RecAdviceProvider {
         self.provider.advance_clock()
     }
 }
+
+impl RecAdviceProvider {
+    // FINALIZATION
+    // --------------------------------------------------------------------------------------------
+    /// Consumes the advice provider and returns an (AdviceInputs, u32, Vec<Felt>, SimpleAdviceMap,
+    /// MerkleStore) tuple.
+    /// The [AdviceInputs] can be used to re-execute the program. The Vec<Felt>, SimpleAdviceMap,
+    /// MerkleStore represent the step, stack, map, and Merkle store of the advice provider at the
+    /// time of finalization.
+    ///
+    /// The returned [AdviceInputs] instance will contain only the non-deterministic inputs which
+    /// were requested during program execution.
+    pub fn finalize(self) -> (AdviceInputs, u32, Vec<Felt>, SimpleAdviceMap, MerkleStore) {
```

Comment:
Similar comment as above: maybe we should remove `step` from the returned tuple?